File: rust/src/decoding/geometry_decoder.rs
use crate::decoding::decoding_utils::DecodingUtils;
use crate::decoding::integer_decoder::IntegerDecoder;
use crate::decoding::stream_metadata_decoder::StreamMetadataDecoder;
use crate::types::compression_types::PhysicalLevelCompressionTechnique;
use crate::types::geometries::{CoordinateType, Geometry, GeometryType};
use crate::types::geometries::coordinate::Coordinate;
use crate::types::geometries::factory::GeometryFactory;
use crate::types::geometries::linearring::LinearRing;
use crate::types::geometry_column::GeometryColumn;
use crate::types::stream_types::{PhysicalStreamType, StreamType_Dictionary, StreamType_VariableSizedItems};

pub struct GeometryDecoder {}
impl GeometryDecoder {
    pub fn decode_geometry_column(data: &[u8], num_streams: usize, offset: &mut usize) -> GeometryColumn {
        let geometry_type_metadata = StreamMetadataDecoder::decode(data, offset);
        let geometry_types = IntegerDecoder::decode_int_stream(data, offset, &geometry_type_metadata, false);


        let mut num_geometries = Vec::new();
        let mut num_parts = Vec::new();
        let mut num_rings = Vec::new();
        let mut vertex_offsets = Some(Vec::new());
        let mut vertex_list = Vec::new();

        for i in 0..num_streams - 1 {
            let geometry_stream_metadata = StreamMetadataDecoder::decode(data, offset);
            match geometry_stream_metadata.stream_metadata.physical_stream_type {
                PhysicalStreamType::LENGTH => {
                    match geometry_stream_metadata.stream_metadata.logical_stream_type.LengthType.unwrap() {
                        StreamType_VariableSizedItems::GEOMETRIES => num_geometries = IntegerDecoder::decode_int_stream(data, offset, &geometry_stream_metadata, false),
                        StreamType_VariableSizedItems::PARTS => num_parts = IntegerDecoder::decode_int_stream(data, offset, &geometry_stream_metadata, false),
                        StreamType_VariableSizedItems::RINGS => num_rings = IntegerDecoder::decode_int_stream(data, offset, &geometry_stream_metadata, false),
                        StreamType_VariableSizedItems::TRIANGLES => panic!("Not implemented yet."),
                        _ => {}
                    }
                },
                PhysicalStreamType::OFFSET => {
                    vertex_offsets = Some(IntegerDecoder::decode_int_stream(data, offset, &geometry_stream_metadata, false));
                },
                PhysicalStreamType::DATA => {
                    if geometry_stream_metadata.stream_metadata.logical_stream_type.DictionaryType.unwrap()
                        == StreamType_Dictionary::VERTEX {
                        if geometry_stream_metadata.stream_metadata.physical_level_technique
                            == PhysicalLevelCompressionTechnique::FAST_PFOR {
                            let vertex_buffer = DecodingUtils::decode_fastpfor_delta_coordinates(
                                data,
                                offset,
                                geometry_stream_metadata.stream_metadata.num_values,
                                geometry_stream_metadata.stream_metadata.byte_length,
                            );
                            vertex_list = vertex_buffer
                                .iter()
                                .map(|i| *i as i32)
                                .collect();
                        } else {
                            vertex_list = IntegerDecoder::decode_int_stream(data, offset, &geometry_stream_metadata, true);
                        }
                    } else {
                        vertex_list = IntegerDecoder::decode_morton_stream(data, offset, &geometry_stream_metadata.morton_stream_metadata.unwrap());
                    }
                }
                PhysicalStreamType::PRESENT => { panic!("Present stream not allowed here"); },
            }
        }

        GeometryColumn {
            num_geometries,
            geometry_types,
            num_parts,
            num_rings,
            vertex_offsets,
            vertex_list,
        }
    }

    fn get_line_string(vertex_buffer: &[i32], start_index: usize, num_vertices: u32, close_line_string: bool) -> Vec<Coordinate> {
        let mut vertices = Vec::with_capacity(if close_line_string { num_vertices + 1 } else { num_vertices } as usize);

        for i in (0..num_vertices as usize * 2).step_by(2) {
            let x = vertex_buffer[start_index + i];
            let y = vertex_buffer[start_index + i + 1];
            vertices.push(Coordinate { x: x as CoordinateType, y: y as CoordinateType });
        }

        if close_line_string {
            vertices.push(vertices[0].clone());
        }

        vertices
    }


    fn decode_dictionary_encoded_line_string(
        vertex_buffer: &[i32],
        vertex_offsets: &[i32],
        vertex_offset: usize,
        num_vertices: u32,
        close_line_string: bool,
    ) -> Vec<Coordinate> {
        let mut vertices = Vec::with_capacity(if close_line_string { num_vertices + 1 } else { num_vertices } as usize);

        for i in 0..num_vertices as usize {
            let offset = vertex_offsets[vertex_offset + i] as usize * 2;
            let x = vertex_buffer[offset];
            let y = vertex_buffer[offset + 1];
            vertices.push(Coordinate { x: x as CoordinateType, y: y as CoordinateType });
        }

        if close_line_string {
            vertices.push(vertices[0].clone());
        }

        vertices
    }

    fn get_linear_ring(
        vertex_buffer: &[i32],
        start_index: usize,
        num_vertices: u32,
    ) -> LinearRing {
        let linear_ring = Self::get_line_string(vertex_buffer, start_index, num_vertices, true);
        GeometryFactory::create_linear_ring(linear_ring)
    }

    fn decode_dictionary_encoded_linear_ring(
        vertex_buffer: &[i32],
        vertex_offsets: &[i32],
        vertex_offset: usize,
        num_vertices: u32,
    ) -> LinearRing {
        let linear_ring = Self::decode_dictionary_encoded_line_string(vertex_buffer, vertex_offsets, vertex_offset, num_vertices, true);
        GeometryFactory::create_linear_ring(linear_ring)
    }

    pub fn decode_geometry(geometry_column: GeometryColumn) -> Vec<Geometry> {
        let mut geometries = Vec::with_capacity(geometry_column.geometry_types.len());

        let mut part_offset_counter = 0;
        let mut ring_offsets_counter = 0;
        let mut geometry_offsets_counter = 0;
        let mut vertex_buffer_offset = 0;
        let mut vertex_offsets_offset = 0;

        let geometry_types = &geometry_column.geometry_types;
        let geometry_offsets = geometry_column.num_geometries;
        let part_offsets = geometry_column.num_parts;
        let ring_offsets = geometry_column.num_rings;
        let vertex_offsets = geometry_column.vertex_offsets.as_ref().map(|v| v.clone()).unwrap_or_default();
        let vertex_buffer = geometry_column.vertex_list.clone();

        let contains_polygon = geometry_column.geometry_types.iter().any(|&g| {
            g == GeometryType::Polygon as i32 || g == GeometryType::MultiPolygon as i32
        });

        for geometry_type in geometry_types {
            match *geometry_type {
                g if g == GeometryType::Point as i32 => {
                    if vertex_offsets.is_empty() {
                        let x = vertex_buffer[vertex_buffer_offset];
                        vertex_buffer_offset += 1;
                        let y = vertex_buffer[vertex_buffer_offset];
                        vertex_buffer_offset += 1;
                        let coordinate = Coordinate { x: x as CoordinateType, y: y as CoordinateType };
                        geometries.push(Geometry::from(GeometryFactory::create_point(coordinate)));
                    } else {
                        let offset = vertex_offsets[vertex_offsets_offset] * 2;
                        vertex_offsets_offset += 1;
                        let x = vertex_buffer[offset as usize];
                        let y = vertex_buffer[offset as usize + 1];
                        let coordinate = Coordinate { x: x as CoordinateType, y: y as CoordinateType };
                        geometries.push(Geometry::from(GeometryFactory::create_point(coordinate)));
                    }
                },
                g if g == GeometryType::MultiPoint as i32 => {
                    let num_points = geometry_offsets[geometry_offsets_counter];
                    geometry_offsets_counter += 1;
                    let mut points = Vec::with_capacity(num_points as usize);
                    if vertex_offsets.is_empty() {
                        for _ in 0..num_points as usize {
                            let x = vertex_buffer[vertex_buffer_offset];
                            vertex_buffer_offset += 1;
                            let y = vertex_buffer[vertex_buffer_offset];
                            vertex_buffer_offset += 1;
                            let coordinate = Coordinate { x: x as CoordinateType, y: y as CoordinateType };
                            points.push(GeometryFactory::create_point(coordinate));
                        }
                        geometries.push(Geometry::from(GeometryFactory::create_multi_point(points)));
                    } else {
                        for _ in 0..num_points {
                            let offset = vertex_offsets[vertex_offsets_offset] * 2;
                            vertex_offsets_offset += 1;
                            let x = vertex_buffer[offset as usize];
                            let y = vertex_buffer[offset as usize + 1];
                            let coordinate = Coordinate { x: x as CoordinateType, y: y as CoordinateType };
                            points.push(GeometryFactory::create_point(coordinate));
                        }
                        geometries.push(Geometry::from(GeometryFactory::create_multi_point(points)));
                    }
                }
                g if g == GeometryType::Linestring as i32 => {
                    let num_vertices = if contains_polygon {
                        ring_offsets_counter += 1;
                        ring_offsets[ring_offsets_counter - 1]
                    } else {
                        part_offset_counter += 1;
                        part_offsets[part_offset_counter - 1]
                    };

                    if vertex_offsets.is_empty() {
                        let vertices = Self::get_line_string(&vertex_buffer, vertex_buffer_offset, num_vertices as u32, false);
                        vertex_buffer_offset += num_vertices as usize * 2;
                        geometries.push(Geometry::from(GeometryFactory::create_line_string(vertices)));
                    } else {
                        let vertices = Self::decode_dictionary_encoded_line_string(
                            &vertex_buffer, &vertex_offsets, vertex_offsets_offset, num_vertices as u32, false);
                        vertex_offsets_offset += num_vertices as usize;

                        geometries.push(Geometry::from(GeometryFactory::create_line_string(vertices)));
                    }
                }
                g if g == GeometryType::Polygon as i32 => {
                    let num_rings = part_offsets[part_offset_counter];
                    part_offset_counter += 1;
                    let mut rings = Vec::with_capacity(num_rings as usize - 1);
                    let mut num_vertices = ring_offsets[ring_offsets_counter];
                    ring_offsets_counter += 1;
                    if vertex_offsets.is_empty() {
                        let shell = Self::get_linear_ring(&vertex_buffer, vertex_buffer_offset, num_vertices as u32);
                        vertex_buffer_offset += num_vertices as usize * 2;
                        for _ in 0..num_rings as usize - 1 {
                            num_vertices = ring_offsets[ring_offsets_counter];
                            ring_offsets_counter += 1;
                            rings.push(
                                Self::get_linear_ring(&vertex_buffer, vertex_buffer_offset, num_vertices as u32));
                            vertex_buffer_offset += num_vertices as usize * 2;
                        }
                        geometries.push(Geometry::from(GeometryFactory::create_polygon(shell, rings)));
                    } else {
                        let shell =
                            Self::decode_dictionary_encoded_linear_ring(&vertex_buffer, &*vertex_offsets, vertex_offsets_offset, num_vertices as u32);
                        vertex_offsets_offset += num_vertices as usize;
                        for _ in 0..num_rings as usize - 1 {
                            num_vertices = ring_offsets[ring_offsets_counter];
                            ring_offsets_counter += 1;
                            rings.push(
                                Self::decode_dictionary_encoded_linear_ring(&vertex_buffer, &vertex_offsets, vertex_offsets_offset, num_vertices as u32));
                            vertex_offsets_offset += num_vertices as usize;
                        }
                        geometries.push(Geometry::from(GeometryFactory::create_polygon(shell, rings)));
                    }
                }
                g if g == GeometryType::MultiLineString as i32 => {
                    let num_line_strings = geometry_offsets[geometry_offsets_counter];
                    geometry_offsets_counter += 1;
                    let mut line_strings = Vec::with_capacity(num_line_strings as usize);
                    if vertex_offsets.is_empty() {
                        for _ in 0..num_line_strings {
                            let num_vertices = if contains_polygon {
                                ring_offsets_counter += 1;
                                ring_offsets[ring_offsets_counter - 1]
                            } else {
                                part_offset_counter += 1;
                                part_offsets[part_offset_counter - 1]
                            };
                            
                            let vertices = Self::get_line_string(&vertex_buffer, vertex_buffer_offset, num_vertices as u32, false);
                            line_strings.push(GeometryFactory::create_line_string(vertices));
                            vertex_buffer_offset += num_vertices as usize * 2;
                        }
                        geometries.push(Geometry::from(GeometryFactory::create_multi_line_string(line_strings)));
                    } else {
                        for _ in 0..num_line_strings {
                            let num_vertices =
                                if contains_polygon {
                                    ring_offsets_counter += 1;
                                    ring_offsets[ring_offsets_counter - 1]
                                } else {
                                    part_offset_counter += 1;
                                    part_offsets[part_offset_counter - 1]
                                };
                            let vertices = 
                                Self::decode_dictionary_encoded_line_string(
                                    &vertex_buffer, &vertex_offsets, vertex_offsets_offset, num_vertices as u32, false);
                            line_strings.push(GeometryFactory::create_line_string(vertices));
                            vertex_offsets_offset += num_vertices as usize;
                        }
                        geometries.push(Geometry::from(GeometryFactory::create_multi_line_string(line_strings)));
                    }
                }
                g if g == GeometryType::MultiPolygon as i32 => {
                    let num_polygons = geometry_offsets[geometry_offsets_counter];
                    geometry_offsets_counter += 1;
                    let mut polygons = Vec::with_capacity(num_polygons as usize);
                    let mut num_vertices;
                    if vertex_offsets.is_empty() {
                        for _ in 0..num_polygons {
                            let num_rings = part_offsets[part_offset_counter];
                            part_offset_counter += 1;
                            let mut rings = Vec::with_capacity(num_rings as usize - 1);
                            num_vertices = ring_offsets[ring_offsets_counter];
                            ring_offsets_counter += 1;
                            let shell =
                                Self::get_linear_ring(
                                    &vertex_buffer, vertex_buffer_offset, num_vertices as u32);
                            vertex_buffer_offset += num_vertices as usize * 2;
                            for j in 0..num_rings as usize - 1 {
                                let num_ring_vertices = ring_offsets[ring_offsets_counter];
                                ring_offsets_counter += 1;
                                rings.push(
                                    Self::get_linear_ring(
                                        &vertex_buffer, vertex_buffer_offset, num_ring_vertices as u32));
                                vertex_buffer_offset += num_ring_vertices as usize * 2;
                            }

                            polygons.push(GeometryFactory::create_polygon(shell, rings));
                        }
                        geometries.push(Geometry::from(GeometryFactory::create_multi_polygon(polygons)));
                    } else {
                        for _ in 0..num_polygons {
                            let num_rings = part_offsets[part_offset_counter];
                            part_offset_counter += 1;
                            let mut rings = Vec::with_capacity(num_rings as usize - 1);
                            num_vertices = ring_offsets[ring_offsets_counter];
                            ring_offsets_counter += 1;
                            let shell =
                                Self::decode_dictionary_encoded_linear_ring(
                                    &vertex_buffer, &vertex_offsets, vertex_offsets_offset, num_vertices as u32);
                            vertex_offsets_offset += num_vertices as usize;
                            for j in 0..num_rings as usize - 1 {
                                num_vertices = ring_offsets[ring_offsets_counter];
                                ring_offsets_counter += 1;
                                rings.push(
                                    Self::decode_dictionary_encoded_linear_ring(
                                        &vertex_buffer, &vertex_offsets, vertex_offsets_offset, num_vertices as u32));
                                vertex_offsets_offset += num_vertices as usize;
                            }
                            polygons.push(GeometryFactory::create_polygon(shell, rings));
                        }
                        geometries.push(Geometry::from(GeometryFactory::create_multi_polygon(polygons)));
                    }
                }
                _ => { panic!("specified geometry type is currently not supported: {}", geometry_type) }
            }
        }
        geometries
    }
}

File: ts/src/decoding/geometryDecoder.ts
import { decodeStreamMetadata, type MortonEncodedStreamMetadata } from "../metadata/tile/streamMetadataDecoder";
import type IntWrapper from "./intWrapper";
import {
    decodeSignedInt32Stream,
    decodeLengthStreamToOffsetBuffer,
    decodeUnsignedConstInt32Stream,
    decodeUnsignedInt32Stream,
    getVectorType,
} from "./integerStreamDecoder";
import { VectorType } from "../vector/vectorType";
import { PhysicalStreamType } from "../metadata/tile/physicalStreamType";
import { LengthType } from "../metadata/tile/lengthType";
import { DictionaryType } from "../metadata/tile/dictionaryType";
import {
    createConstGeometryVector,
    createMortonEncodedConstGeometryVector,
} from "../vector/geometry/constGeometryVector";
import { createFlatGeometryVector, createFlatGeometryVectorMortonEncoded } from "../vector/geometry/flatGeometryVector";
import { OffsetType } from "../metadata/tile/offsetType";
import { createConstGpuVector } from "../vector/geometry/constGpuVector";
import { createFlatGpuVector } from "../vector/geometry/flatGpuVector";
import type { GeometryVector, MortonSettings } from "../vector/geometry/geometryVector";
import type { GpuVector } from "../vector/geometry/gpuVector";
import type GeometryScaling from "./geometryScaling";

// TODO: get rid of numFeatures parameter
export function decodeGeometryColumn(
    tile: Uint8Array,
    numStreams: number,
    offset: IntWrapper,
    numFeatures: number,
    scalingData?: GeometryScaling,
): GeometryVector | GpuVector {
    const geometryTypeMetadata = decodeStreamMetadata(tile, offset);
    const geometryTypesVectorType = getVectorType(geometryTypeMetadata, numFeatures, tile, offset);

    let vertexOffsets: Uint32Array | undefined;
    let vertexBuffer: Int32Array | Uint32Array | undefined;
    let mortonSettings: MortonSettings | undefined;
    let indexBuffer: Uint32Array | undefined;

    if (geometryTypesVectorType === VectorType.CONST) {
        /* All geometries in the column have the same geometry type */
        const geometryType = decodeUnsignedConstInt32Stream(tile, offset, geometryTypeMetadata);

        // Variables for const geometry path (directly decoded as offsets)
        let geometryOffsets: Uint32Array | undefined;
        let partOffsets: Uint32Array | undefined;
        let ringOffsets: Uint32Array | undefined;
        //TODO: use geometryOffsets for that? -> but then tessellated polygons can't be used with normal polygons
        // in one FeatureTable?
        let triangleOffsets: Uint32Array | undefined;

        for (let i = 0; i < numStreams - 1; i++) {
            const geometryStreamMetadata = decodeStreamMetadata(tile, offset);
            switch (geometryStreamMetadata.physicalStreamType) {
                case PhysicalStreamType.LENGTH:
                    switch (geometryStreamMetadata.logicalStreamType.lengthType) {
                        case LengthType.GEOMETRIES:
                            geometryOffsets = decodeLengthStreamToOffsetBuffer(tile, offset, geometryStreamMetadata);
                            break;
                        case LengthType.PARTS:
                            partOffsets = decodeLengthStreamToOffsetBuffer(tile, offset, geometryStreamMetadata);
                            break;
                        case LengthType.RINGS:
                            ringOffsets = decodeLengthStreamToOffsetBuffer(tile, offset, geometryStreamMetadata);
                            break;
                        case LengthType.TRIANGLES:
                            triangleOffsets = decodeLengthStreamToOffsetBuffer(tile, offset, geometryStreamMetadata);
                    }
                    break;
                case PhysicalStreamType.OFFSET: {
                    switch (geometryStreamMetadata.logicalStreamType.offsetType) {
                        case OffsetType.VERTEX:
                            vertexOffsets = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata);
                            break;
                        case OffsetType.INDEX:
                            indexBuffer = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata);
                            break;
                    }
                    break;
                }
                case PhysicalStreamType.DATA: {
                    if (DictionaryType.VERTEX === geometryStreamMetadata.logicalStreamType.dictionaryType) {
                        vertexBuffer = decodeSignedInt32Stream(tile, offset, geometryStreamMetadata, scalingData);
                    } else {
                        const mortonMetadata = geometryStreamMetadata as MortonEncodedStreamMetadata;
                        mortonSettings = {
                            numBits: mortonMetadata.numBits,
                            coordinateShift: mortonMetadata.coordinateShift,
                        };
                        vertexBuffer = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata, scalingData);
                    }
                    break;
                }
            }
        }

        if (indexBuffer) {
            if (geometryOffsets !== undefined || partOffsets !== undefined) {
                /* Case when the indices of a Polygon outline are encoded in the tile */
                const topologyVector = { geometryOffsets, partOffsets, ringOffsets };
                return createConstGpuVector(
                    numFeatures,
                    geometryType,
                    triangleOffsets,
                    indexBuffer,
                    vertexBuffer,
                    topologyVector,
                );
            }

            /* Case when the no Polygon outlines are encoded in the tile */
            return createConstGpuVector(numFeatures, geometryType, triangleOffsets, indexBuffer, vertexBuffer);
        }

        return mortonSettings === undefined
            ? /* Currently only 2D coordinates (Vec2) are implemented in the encoder  */
              createConstGeometryVector(
                  numFeatures,
                  geometryType,
                  { geometryOffsets, partOffsets, ringOffsets },
                  vertexOffsets,
                  vertexBuffer,
              )
            : createMortonEncodedConstGeometryVector(
                  numFeatures,
                  geometryType,
                  { geometryOffsets, partOffsets, ringOffsets },
                  vertexOffsets,
                  vertexBuffer,
                  mortonSettings,
              );
    }

    /* Different geometry types are mixed in the geometry column */
    const geometryTypeVector = decodeUnsignedInt32Stream(tile, offset, geometryTypeMetadata);

    // Variables for flat geometry path (decoded as lengths, then converted to offsets)
    let geometryLengths: Uint32Array | undefined;
    let partLengths: Uint32Array | undefined;
    let ringLengths: Uint32Array | undefined;
    //TODO: use geometryOffsets for that? -> but then tessellated polygons can't be used with normal polygons
    // in one FeatureTable?
    let triangleOffsets: Uint32Array | undefined;

    for (let i = 0; i < numStreams - 1; i++) {
        const geometryStreamMetadata = decodeStreamMetadata(tile, offset);
        switch (geometryStreamMetadata.physicalStreamType) {
            case PhysicalStreamType.LENGTH:
                switch (geometryStreamMetadata.logicalStreamType.lengthType) {
                    case LengthType.GEOMETRIES:
                        geometryLengths = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata);
                        break;
                    case LengthType.PARTS:
                        partLengths = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata);
                        break;
                    case LengthType.RINGS:
                        ringLengths = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata);
                        break;
                    case LengthType.TRIANGLES:
                        triangleOffsets = decodeLengthStreamToOffsetBuffer(tile, offset, geometryStreamMetadata);
                }
                break;
            case PhysicalStreamType.OFFSET:
                switch (geometryStreamMetadata.logicalStreamType.offsetType) {
                    case OffsetType.VERTEX:
                        vertexOffsets = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata);
                        break;
                    case OffsetType.INDEX:
                        indexBuffer = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata);
                        break;
                }
                break;
            case PhysicalStreamType.DATA:
                if (DictionaryType.VERTEX === geometryStreamMetadata.logicalStreamType.dictionaryType) {
                    vertexBuffer = decodeSignedInt32Stream(tile, offset, geometryStreamMetadata, scalingData);
                } else {
                    const mortonMetadata = geometryStreamMetadata as MortonEncodedStreamMetadata;
                    mortonSettings = {
                        numBits: mortonMetadata.numBits,
                        coordinateShift: mortonMetadata.coordinateShift,
                    };
                    vertexBuffer = decodeUnsignedInt32Stream(tile, offset, geometryStreamMetadata, scalingData);
                }
                break;
        }
    }

    // TODO: refactor the following instructions -> decode in one pass for performance reasons
    /* Calculate the offsets from the length buffer for util access */
    let geometryOffsets: Uint32Array | undefined;
    let partOffsets: Uint32Array | undefined;
    let ringOffsets: Uint32Array | undefined;

    if (geometryLengths) {
        geometryOffsets = decodeRootLengthStream(geometryTypeVector, geometryLengths, 2);
        if (partLengths && ringLengths) {
            partOffsets = decodeLevel1LengthStream(geometryTypeVector, geometryOffsets, partLengths, false);
            ringOffsets = decodeLevel2LengthStream(geometryTypeVector, geometryOffsets, partOffsets, ringLengths);
        } else if (partLengths) {
            partOffsets = decodeLevel1WithoutRingBufferLengthStream(geometryTypeVector, geometryOffsets, partLengths);
        }
    } else if (partLengths && ringLengths) {
        partOffsets = decodeRootLengthStream(geometryTypeVector, partLengths, 1);
        ringOffsets = decodeLevel1LengthStream(geometryTypeVector, partOffsets, ringLengths, true);
    } else if (partLengths) {
        partOffsets = decodeRootLengthStream(geometryTypeVector, partLengths, 0);
    }

    if (indexBuffer && !partOffsets) {
        /* Case when the indices of a Polygon outline are not encoded in the data so no
         *  topology data are present in the tile */
        return createFlatGpuVector(geometryTypeVector, triangleOffsets, indexBuffer, vertexBuffer);
    }

    if (indexBuffer) {
        /* Case when the indices of a Polygon outline are encoded in the tile */
        return createFlatGpuVector(geometryTypeVector, triangleOffsets, indexBuffer, vertexBuffer, {
            geometryOffsets,
            partOffsets,
            ringOffsets,
        });
    }

    return mortonSettings === undefined /* Currently only 2D coordinates (Vec2) are implemented in the encoder  */
        ? createFlatGeometryVector(
              geometryTypeVector,
              { geometryOffsets, partOffsets, ringOffsets },
              vertexOffsets,
              vertexBuffer,
          )
        : createFlatGeometryVectorMortonEncoded(
              geometryTypeVector,
              { geometryOffsets, partOffsets, ringOffsets },
              vertexOffsets,
              vertexBuffer,
              mortonSettings,
          );
}

/*
 * Handle the parsing of the different topology length buffers separate not generic to reduce the
 * branching and improve the performance
 */
function decodeRootLengthStream(
    geometryTypes: Uint32Array,
    rootLengthStream: Uint32Array,
    bufferId: number,
): Uint32Array {
    const rootBufferOffsets = new Uint32Array(geometryTypes.length + 1);
    let previousOffset = 0;
    rootBufferOffsets[0] = previousOffset;
    let rootLengthCounter = 0;
    for (let i = 0; i < geometryTypes.length; i++) {
        /* Test if the geometry has and entry in the root buffer
         * BufferId: 2 GeometryOffsets -> MultiPolygon, MultiLineString, MultiPoint
         * BufferId: 1 PartOffsets -> Polygon
         * BufferId: 0 PartOffsets, RingOffsets -> LineString
         * */
        previousOffset = rootBufferOffsets[i + 1] =
            previousOffset + (geometryTypes[i] > bufferId ? rootLengthStream[rootLengthCounter++] : 1);
    }

    return rootBufferOffsets;
}

function decodeLevel1LengthStream(
    geometryTypes: Uint32Array,
    rootOffsetBuffer: Uint32Array,
    level1LengthBuffer: Uint32Array,
    isLineStringPresent: boolean,
): Uint32Array {
    const level1BufferOffsets = new Uint32Array(rootOffsetBuffer[rootOffsetBuffer.length - 1] + 1);
    let previousOffset = 0;
    level1BufferOffsets[0] = previousOffset;
    let level1BufferCounter = 1;
    let level1LengthBufferCounter = 0;
    for (let i = 0; i < geometryTypes.length; i++) {
        const geometryType = geometryTypes[i];
        const numGeometries = rootOffsetBuffer[i + 1] - rootOffsetBuffer[i];
        if (
            geometryType === 5 ||
            geometryType === 2 ||
            (isLineStringPresent && (geometryType === 4 || geometryType === 1))
        ) {
            /* For MultiPolygon, Polygon and in some cases for MultiLineString and LineString
             * a value in the level1LengthBuffer exists */
            for (let j = 0; j < numGeometries; j++) {
                previousOffset = level1BufferOffsets[level1BufferCounter++] =
                    previousOffset + level1LengthBuffer[level1LengthBufferCounter++];
            }
        } else {
            /* For MultiPoint and Point and in some cases for MultiLineString and LineString no value in the
             * level1LengthBuffer exists */
            for (let j = 0; j < numGeometries; j++) {
                level1BufferOffsets[level1BufferCounter++] = ++previousOffset;
            }
        }
    }

    return level1BufferOffsets;
}

/*
 * Case where no ring buffer exists so no MultiPolygon or Polygon geometry is part of the buffer
 */
function decodeLevel1WithoutRingBufferLengthStream(
    geometryTypes: Uint32Array,
    rootOffsetBuffer: Uint32Array,
    level1LengthBuffer: Uint32Array,
): Uint32Array {
    const level1BufferOffsets = new Uint32Array(rootOffsetBuffer[rootOffsetBuffer.length - 1] + 1);
    let previousOffset = 0;
    level1BufferOffsets[0] = previousOffset;
    let level1OffsetBufferCounter = 1;
    let level1LengthCounter = 0;
    for (let i = 0; i < geometryTypes.length; i++) {
        const geometryType = geometryTypes[i];
        const numGeometries = rootOffsetBuffer[i + 1] - rootOffsetBuffer[i];
        if (geometryType === 4 || geometryType === 1) {
            /* For MultiLineString and LineString a value in the level1LengthBuffer exists */
            for (let j = 0; j < numGeometries; j++) {
                previousOffset = level1BufferOffsets[level1OffsetBufferCounter++] =
                    previousOffset + level1LengthBuffer[level1LengthCounter++];
            }
        } else {
            /* For MultiPoint and Point no value in level1LengthBuffer exists */
            for (let j = 0; j < numGeometries; j++) {
                level1BufferOffsets[level1OffsetBufferCounter++] = ++previousOffset;
            }
        }
    }

    return level1BufferOffsets;
}

function decodeLevel2LengthStream(
    geometryTypes: Uint32Array,
    rootOffsetBuffer: Uint32Array,
    level1OffsetBuffer: Uint32Array,
    level2LengthBuffer: Uint32Array,
): Uint32Array {
    const level2BufferOffsets = new Uint32Array(level1OffsetBuffer[level1OffsetBuffer.length - 1] + 1);
    let previousOffset = 0;
    level2BufferOffsets[0] = previousOffset;
    let level1OffsetBufferCounter = 1;
    let level2OffsetBufferCounter = 1;
    let level2LengthBufferCounter = 0;
    for (let i = 0; i < geometryTypes.length; i++) {
        const geometryType = geometryTypes[i];
        const numGeometries = rootOffsetBuffer[i + 1] - rootOffsetBuffer[i];
        if (geometryType !== 0 && geometryType !== 3) {
            /* For MultiPolygon, MultiLineString, Polygon and LineString a value in level2LengthBuffer
             * exists */
            for (let j = 0; j < numGeometries; j++) {
                const numParts =
                    level1OffsetBuffer[level1OffsetBufferCounter] - level1OffsetBuffer[level1OffsetBufferCounter - 1];
                level1OffsetBufferCounter++;
                for (let k = 0; k < numParts; k++) {
                    previousOffset = level2BufferOffsets[level2OffsetBufferCounter++] =
                        previousOffset + level2LengthBuffer[level2LengthBufferCounter++];
                }
            }
        } else {
            /* For MultiPoint and Point no value in level2LengthBuffer exists */
            for (let j = 0; j < numGeometries; j++) {
                level2BufferOffsets[level2OffsetBufferCounter++] = ++previousOffset;
                level1OffsetBufferCounter++;
            }
        }
    }

    return level2BufferOffsets;
}
